### ISSUE TITLE
derive-new errors

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#66525]
So @wycats pointed out to me that some recent changes in name resolution seem to have broken `derive_new`. They pointed me at [this commit from nushell](https://github.com/nushell/nushell/commit/372f6c16b31dab03af0f6adc5e050884f2d854c2) dealing with the fallout -- @jonathandturner maybe you can leave a few notes as to what errors you were seeing before this commit?

cc @petrochenkov -- could this be the recent changes around name resolution of helper attributes in macros? It seems the relevant code is stuff like:

```rust
#[derive(new)]
struct Foo {
    #[new(value = "Arc::new(22)")]
    bar: Arc<i32>
}
```

**Comment by nikomatsakis** [comment]
(Nominated for @rust-lang/lang discussion around the breakage policy/implications here.)

**Comment by petrochenkov** [comment]
Duplicate of https://github.com/rust-lang/rust/issues/66508, I'll prepare a fix.

**Comment by petrochenkov** [comment]
>fix

Or rather a "tiebreaker" for the ambiguity, since both names are indeed in scope and create ambiguity, which is correctly detected after #64694.

**Comment by petrochenkov** [comment]
Fixed in https://github.com/rust-lang/rust/pull/66529.
